src/engine/blocks.js

```diff
@@ -1080,25 +1128,33 @@ class Blocks {
                 xmlString += '</value>';
             }
         }
-        // Add any fields on this block.
-        for (const field in block.fields) {
-            if (!block.fields.hasOwnProperty(field)) continue;
-            const blockField = block.fields[field];
-            xmlString += `<field name="${blockField.name}"`;
-            const fieldId = blockField.id;
-            if (fieldId) {
-                xmlString += ` id="${fieldId}"`;
-            }
-            const varType = blockField.variableType;
-            if (typeof varType === 'string') {
-                xmlString += ` variabletype="${varType}"`;
-            }
-            let value = blockField.value;
-            if (typeof value === 'string') {
-                value = xmlEscape(blockField.value);
+
+        // Add fields to the block, but only if the block itself
+        // doesn't have a mutation with a blockinfo.
+        // This is a dynamic extension block which will be laid out using
+        // it's blockInfo mutation
+        if (!(block.mutation && block.mutation.blockInfo)) {
```

Comment:
I am putting the code for generating the XML for a field inside of this check that the block is not dynamic so that we don't emit XML that has both a dynamic extension block with a mutation determining how to lay it out, and also conflicting field information.

There's a trade off here, because this is working towards making the field value more obsolete (if we accomplish our goal of switching all blocks to be dynamic extension blocks, I think the `fields` value of the VM JSON representation of blocks will be completely obsolete).

However, if we think that is too risky, the alternative is to keep this code dependent on scratch-blocks, and react to the scratch-blocks `var_rename` event emitted by the variables modal. I have an alternative set of changes stashed away that handle keeping the block field values in sync in the VM (as we do currently with the regular variable blocks right now) by adding code that updates the blockInfo to the pre-existing `updateBlocksAfterVarRename` function that updates all the field values and ids for all the relevant blocks referencing the renamed variable. In an earlier comment, I mentioned that this code would be obsolete post-extensionification, taking this alternative would make that statement false.

I think the latter alternative is a little more confusing, because it makes the blocks unnecessarily dependent on both the blockInfo mutation and the "fields" part of the VM JSON representation. I think this will ultimately lead to confusion down the line (post-extensionification)....

Reply:
Thanks for the thorough explanation here! I agree with you -- unless we run into some sort of Big Problem I think it makes sense to work toward obsoleting the `fields` value in the VM JSON.